File: pokeop/api/schemas/winning_paths.py
"""定义按配置分组的胜利路径 Top-K HTTP 请求与响应合同。"""

from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from pokeop.api.schemas.battle_exploration import (
    ExactProbabilityResponse,
    ExplorationPathStepResponse,
)
from pokeop.application.use_cases.winning_paths import WinningPathGroupsResult


class _AttributeResponseModel(BaseModel):
    """允许响应模型直接读取 application 不可变 dataclass 字段。"""

    model_config = ConfigDict(from_attributes=True)


class ListWinningPathGroupsRequest(BaseModel):
    """请求一个固定 graph 配置下指定获胜侧的分页路径组。"""

    calculation_revision: str = Field(min_length=1)
    winner: str = Field(pattern="^(attacker|defender)$")
    limit: int = Field(default=10, ge=1, le=100)
    cursor: str | None = None
    sort: str = Field(default="shortest-high-probability")


class WinningPathConfigurationSideResponse(_AttributeResponseModel):
    """返回配置身份中的一方 Pokémon、特性、道具和招式组。"""

    pokemon_id: int
    name: str
    level: int
    ability_identifier: str
    item_identifier: str
    move_ids: list[int]


class WinningPathConfigurationResponse(_AttributeResponseModel):
    """返回不会跨 graph 配置归并的稳定双方配置摘要。"""

    configuration_key: str
    attacker: WinningPathConfigurationSideResponse
    defender: WinningPathConfigurationSideResponse


class JointActionAlternativeResponse(_AttributeResponseModel):
    """返回一条原始解释中的双方联合行动。"""

    attacker_move_id: int | None
    defender_move_id: int | None
    attacker_source_identifier: str | None
    defender_source_identifier: str | None


class JointActionStepResponse(_AttributeResponseModel):
    """返回一个回合的联合行动主步骤及歧义候选。"""

    turn_number: int
    attacker_move_id: int | None
    defender_move_id: int | None
    attacker_source_identifier: str | None
    defender_source_identifier: str | None
    ambiguous: bool
    alternatives: list[JointActionAlternativeResponse]


class WinningPathKeyEventResponse(_AttributeResponseModel):
    """返回命中失败、行动阻断、机制触发、状态或濒死关键事件。"""

    kind: str
    actor: str | None
    target: str | None
    move_id: int | None
    source_identifier: str | None


class WinningPathGroupResponse(_AttributeResponseModel):
    """返回同一配置和行动序列归并后的一个胜利路径组。"""

    path_key: str
    terminal_turn: int
    probability: ExactProbabilityResponse
    raw_path_count: int
    raw_history_count_estimate: int
    terminal_reasons: list[str]
    terminal_node_ids: list[int]
    actions: list[JointActionStepResponse]
    representative_path: list[ExplorationPathStepResponse]
    damage_values: list[int]
    attacker_remaining_hp_values: list[int]
    defender_remaining_hp_values: list[int]
    key_events: list[WinningPathKeyEventResponse]


class WinningPathPrefixNodeResponse(_AttributeResponseModel):
    """返回当前页路径组构造的递归行动前缀树节点。"""

    prefix_key: str
    depth: int
    action: JointActionStepResponse | None
    probability: ExactProbabilityResponse
    raw_path_count: int
    terminal_path_keys: list[str]
    children: list["WinningPathPrefixNodeResponse"]


class WinningPathCycleReferenceResponse(_AttributeResponseModel):
    """返回被压缩为有限引用的回边或重复节点。"""

    source_node_id: int
    edge_id: int
    target_node_id: int
    prefix_depth: int
    component_id: int | None


class WinningPathGroupsResponse(_AttributeResponseModel):
    """返回胜利路径分页窗口、概率覆盖、前缀树与完整性语义。"""

    graph_id: str
    calculation_revision: str
    winner: str
    sort: str
    configuration: WinningPathConfigurationResponse
    winner_probability: ExactProbabilityResponse | None
    returned_probability: ExactProbabilityResponse
    returned_coverage: ExactProbabilityResponse | None
    path_groups: list[WinningPathGroupResponse]
    prefix_tree: WinningPathPrefixNodeResponse
    cycle_references: list[WinningPathCycleReferenceResponse]
    next_cursor: str | None
    has_more: bool
    query_complete: bool
    traversal_truncated: bool


def winning_path_groups_response(
    result: WinningPathGroupsResult,
) -> WinningPathGroupsResponse:
    """把 application 查询结果转换为不泄漏图对象的 HTTP DTO。

    Args:
        result: 已完成 graph、分页游标和概率校验的 application 结果。

    Returns:
        可由 FastAPI 直接序列化的稳定胜利路径响应。
    """
    return WinningPathGroupsResponse.model_validate(result)


__all__ = [
    "ListWinningPathGroupsRequest",
    "WinningPathGroupsResponse",
    "winning_path_groups_response",
]

File: web/src/api/winningPaths.ts
import type { ExplorationPathStepResult, ProbabilityResult } from './inference';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? '/api/v1';

export type WinningPathWinner = 'attacker' | 'defender';

export interface WinningPathConfigurationSideResult {
  pokemon_id: number;
  name: string;
  level: number;
  ability_identifier: string;
  item_identifier: string;
  move_ids: number[];
}

export interface WinningPathConfigurationResult {
  configuration_key: string;
  attacker: WinningPathConfigurationSideResult;
  defender: WinningPathConfigurationSideResult;
}

export interface JointActionAlternativeResult {
  attacker_move_id: number | null;
  defender_move_id: number | null;
  attacker_source_identifier: string | null;
  defender_source_identifier: string | null;
}

export interface JointActionStepResult {
  turn_number: number;
  attacker_move_id: number | null;
  defender_move_id: number | null;
  attacker_source_identifier: string | null;
  defender_source_identifier: string | null;
  ambiguous: boolean;
  alternatives: JointActionAlternativeResult[];
}

export interface WinningPathKeyEventResult {
  kind: string;
  actor: string | null;
  target: string | null;
  move_id: number | null;
  source_identifier: string | null;
}

export interface WinningPathGroupResult {
  path_key: string;
  terminal_turn: number;
  probability: ProbabilityResult;
  raw_path_count: number;
  raw_history_count_estimate: number;
  terminal_reasons: string[];
  terminal_node_ids: number[];
  actions: JointActionStepResult[];
  representative_path: ExplorationPathStepResult[];
  damage_values: number[];
  attacker_remaining_hp_values: number[];
  defender_remaining_hp_values: number[];
  key_events: WinningPathKeyEventResult[];
}

export interface WinningPathPrefixNodeResult {
  prefix_key: string;
  depth: number;
  action: JointActionStepResult | null;
  probability: ProbabilityResult;
  raw_path_count: number;
  terminal_path_keys: string[];
  children: WinningPathPrefixNodeResult[];
}

export interface WinningPathCycleReferenceResult {
  source_node_id: number;
  edge_id: number;
  target_node_id: number;
  prefix_depth: number;
  component_id: number | null;
}

export interface WinningPathGroupsResult {
  graph_id: string;
  calculation_revision: string;
  winner: WinningPathWinner;
  sort: 'shortest-high-probability';
  configuration: WinningPathConfigurationResult;
  winner_probability: ProbabilityResult | null;
  returned_probability: ProbabilityResult;
  returned_coverage: ProbabilityResult | null;
  path_groups: WinningPathGroupResult[];
  prefix_tree: WinningPathPrefixNodeResult;
  cycle_references: WinningPathCycleReferenceResult[];
  next_cursor: string | null;
  has_more: boolean;
  query_complete: boolean;
  traversal_truncated: boolean;
}

interface WinningPathErrorPayload {
  detail?: string | { message?: string };
}

/**
 * 查询固定配置下一个获胜侧的分页行动路径组。
 *
 * @param graphId 首次推演返回的 graph ID。
 * @param calculationRevision 首次推演返回的计算语义版本。
 * @param winner 需要解释的绝对获胜侧。
 * @param limit 本页最多返回的路径组数量。
 * @param cursor 上一页返回的不透明游标；首页传 null。
 * @returns 当前页 Top-K、行动前缀树、精确概率覆盖与循环引用。
 * @throws Error 当后端返回非 2xx 状态时抛出用户可读错误。
 */
export async function listWinningPathGroups(
  graphId: string,
  calculationRevision: string,
  winner: WinningPathWinner,
  limit = 10,
  cursor: string | null = null,
): Promise<WinningPathGroupsResult> {
  const response = await fetch(
    `${API_BASE_URL}/winning_paths/graphs/${encodeURIComponent(graphId)}`,
    {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        calculation_revision: calculationRevision,
        winner,
        limit,
        cursor,
        sort: 'shortest-high-probability',
      }),
    },
  );
  const payload = (await response.json()) as WinningPathGroupsResult & WinningPathErrorPayload;
  if (!response.ok) {
    const detail = payload.detail;
    const message =
      typeof detail === 'string'
        ? detail
        : detail?.message ?? `胜利路径查询失败：HTTP ${response.status}`;
    throw new Error(message);
  }
  return payload;
}
